soc/nordic/common/dmm.c

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, void ** buffer_out)
+{
+    // Get memory region that specified device can perform DMA transfers from, using devicetree
+
+    // Check if provided user buffer is already in correct memory region
+    // If yes, assign buffer_out to user_buffer
+    // If no:
+    // - dynamically allocate buffer in correct memory region, using for example mem_attr_heap_alloc()
+    // - copy user buffer contents into allocated buffer
+
+    // Return error if dynamic allocation fails
+
+    // Check if device memory region is cacheable
+    // If yes, writeback all cache lines associated with output buffer (either user or allocated)
+    // If no, no action is needed
+}
+
+int dmm_dma_buffer_out_release(void const * buffer_out)
+{
+    // Get memory region that specified device can perform DMA transfers from, using devicetree
+
+    // Check if output buffer is contained within memory area managed by dynamic memory allocator
+    // If yes, free the buffer
+    // If no, no action is needed
+}
+
+int dmm_dma_buffer_in_prepare(const struct device *dev, void const * user_buffer, void ** buffer_in)
+{
+    // Get memory region that specified device can perform DMA transfers to, using devicetree
+
+    // Check if provided user buffer is already in correct memory region
+    // If yes, assign buffer_in to user_buffer
+    // If no, dynamically allocate buffer in correct memory region, using for example mem_attr_heap_alloc()
+
+    // Return error if dynamic allocation fails
+}
+
+int dmm_dma_buffer_in_release(void const * user_buffer, void const * buffer_in)
+{
+    // Get memory region that specified device can perform DMA transfers to, using devicetree
+
+    // Check if input buffer is contained within memory area managed by dynamic memory allocator
+    // If yes, free the buffer
```

Comment:
I wouldn't free the buffer until its content is copied to the user buffer.

Reply:
fixed - changed the order of the operations